tests/zfs-tests/tests/functional/fault/auto_spare_001_pos.ksh

```diff
@@ -0,0 +1,145 @@
+# See the License for the specific language governing permissions
+# and limitations under the License.
+#
+# When distributing Covered Code, include this CDDL HEADER in each
+# file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+# If applicable, add the following below this CDDL HEADER, with the
+# fields enclosed by brackets "[]" replaced with your own identifying
+# information: Portions Copyright [yyyy] [name of copyright owner]
+#
+# CDDL HEADER END
+#
+#
+# Copyright (c) 2017 by Intel Corporation. All rights reserved.
+#
+
+. $STF_SUITE/include/libtest.shlib
+. $STF_SUITE/tests/functional/fault/fault.cfg
+
+#
+# DESCRIPTION:
+# Testing Fault Management Agent ZED Logic - Automated Auto-Spare Test.
+#
+# STRATEGY:
+# 1. Create a raidz pool with hot spares
+# 2. Write multiple files to the pool
+# 3. Export the pool to clear the cache
+# 4. Import the pool
+# 5. Configure zinject to cause read io errors on the first drive in the pool
+# 6. Read data from the pool to generate io errors
+# 7. ZED detects the io errors and faults the drive
+# 8. ZED using the retire agent swaps in a spare for the faulted drive
+# 9. poll for a spare event change
+# 10. Remove zinject handlers
+# 11. Clear events
+# 12. Clear the status of the faulted drive
+# 13. ZED recognizes the resilver and when it is complete removes the spare
+# 14. poll for the resilver_finish event
+#
+
+verify_runnable "both"
+
+if ! is_physical_device $DISKS; then
+	log_unsupported "Unsupported disks for this test."
+fi
+
+function setup
+{
+        cleanup_devices $DISK1
+        partition_disk $SIZE $DISK1 3
+	cleanup_devices $DISK2
+	partition_disk $SIZE $DISK2 3
+	cleanup_devices $DISK3
+	partition_disk $SIZE $DISK3 2
+}
+
+function cleanup
+{
+	log_must zinject -c all
+	poolexists $TESTPOOL && destroy_pool $TESTPOOL
+}
+
+log_assert "Testing automated auto-spare FMA test"
+
+log_onexit cleanup
+
+# Clear disk labels
+zpool labelclear -f $DISK1
+zpool labelclear -f $DISK2
+zpool labelclear -f $DISK3
+
+setup
+
+log_must zpool create -f $TESTPOOL raidz ${RAIDZ_DISK0} ${RAIDZ_DISK1} spare ${RAIDZ_SPARES}
+log_must zfs set recordsize=1MB $TESTPOOL
+# Add some data to the pool
+for x in `seq 1 20`
+do
+	log_must mkfile $FSIZE /$TESTPOOL/data${x}
+done
+
+guid=$(get_config $TESTPOOL pool_guid)
+# Clear the caches by exporting and importing
+log_must zpool export $TESTPOOL
+log_must zpool import -s $guid
+
+# Clear zpool events
+log_must zpool events -c $TESTPOOL
+
+# Generate faults so ZED faults the drive
+log_must zinject -d ${FAULT_DRIVE} -e io -T read $TESTPOOL
+for f in /$TESTPOOL/*
+do
+	cat $f > /dev/null
+done
+
+log_note "Delay for ZED auto-spare"
+typeset -i timeout=0
+while true; do
+	if ((timeout == $MAXTIMEOUT)); then
+		log_fail "Timeout occured"
+	fi
+	((timeout++))
+	sleep 1
+	zpool events $TESTPOOL | egrep sysevent.fs.zfs.vdev_spare \
+	    > /dev/null
+	if (($? == 0)); then
+		log_note "Auto-spare should be complete"
```

Comment:
You can use the existing `wait_replacing` and `check_state` functions to verify that the the rebuild is done and the device are in the correct state.